src/python/grpcio_observability/grpc_observability/server_call_tracer.cc

```diff
@@ -0,0 +1,247 @@
+    grpc_metadata_batch* recv_initial_metadata) {
+  ServerO11yMetadata som;
+  GetO11yMetadata(recv_initial_metadata, &som);
+  path_ = std::move(som.path);
+  method_ = GetMethod(path_);
+  auto tracing_enabled = PythonCensusTracingEnabled();
+  GenerateServerContext(
+      tracing_enabled ? som.tracing_slice.as_string_view() : "",
+      absl::StrCat("Recv.", method_), &context_);
+  if (PythonCensusStatsEnabled()) {
+    context_.Labels().emplace_back(kServerMethod, std::string(method_));
+    RecordIntMetric(kRpcServerStartedRpcsMeasureName, 1, context_.Labels());
+  }
+}
+
+void PythonOpenCensusServerCallTracer::RecordSendTrailingMetadata(
+    grpc_metadata_batch* send_trailing_metadata) {
+  // We need to record the time when the trailing metadata was sent to
+  // mark the completeness of the request.
+  elapsed_time_ = absl::Now() - start_time_;
+  if (PythonCensusStatsEnabled() && send_trailing_metadata != nullptr) {
+    size_t len = ServerStatsSerialize(absl::ToInt64Nanoseconds(elapsed_time_),
+                                      stats_buf_, kMaxServerStatsLen);
+    if (len > 0) {
+      send_trailing_metadata->Set(
+          grpc_core::GrpcServerStatsBinMetadata(),
+          grpc_core::Slice::FromCopiedBuffer(stats_buf_, len));
+    }
+  }
+}
+
+void PythonOpenCensusServerCallTracer::RecordEnd(
+    const grpc_call_final_info* final_info) {
+  if (PythonCensusStatsEnabled()) {
+    const uint64_t request_size = GetOutgoingDataSize(final_info);
+    const uint64_t response_size = GetIncomingDataSize(final_info);
+    double elapsed_time_ms = absl::ToDoubleMilliseconds(elapsed_time_);
+    context_.Labels().emplace_back(kServerMethod, std::string(method_));
+    context_.Labels().emplace_back(
+        kServerStatus,
+        std::string(StatusCodeToString(final_info->final_status)));
+    RecordDoubleMetric(kRpcServerSentBytesPerRpcMeasureName,
+                       static_cast<double>(response_size), context_.Labels());
+    RecordDoubleMetric(kRpcServerReceivedBytesPerRpcMeasureName,
+                       static_cast<double>(request_size), context_.Labels());
+    RecordDoubleMetric(kRpcServerServerLatencyMeasureName, elapsed_time_ms,
+                       context_.Labels());
+    RecordIntMetric(kRpcServerSentMessagesPerRpcMeasureName,
+                    sent_message_count_, context_.Labels());
+    RecordIntMetric(kRpcServerReceivedMessagesPerRpcMeasureName,
+                    recv_message_count_, context_.Labels());
+  }
+  if (PythonCensusTracingEnabled()) {
+    context_.EndSpan();
+    if (IsSampled()) {
+      RecordSpan(context_.Span().ToCensusData());
+    }
+  }
+
+  // After RecordEnd, Core will make no further usage of this ServerCallTracer,
+  // so we are free it here.
+  delete this;
+}
+
+//
+// PythonOpenCensusServerCallTracerFactory
+//
+
+grpc_core::ServerCallTracer*
+PythonOpenCensusServerCallTracerFactory::CreateNewServerCallTracer(
+    grpc_core::Arena* arena) {
```

Comment:
Probably want an explicit `(void)arena;` to indicate that we're intentionally ignoring this argument.